attention_and_transformers.ipynb

```diff
@@ -0,0 +1,4261 @@
+      "metadata": {
+        "id": "u3WZyV_s54_R"
+      },
+      "source": [
+        "If everything is correct, then if we run the code below, everything should run without any issues."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "LbRxqe_lfRIq"
+      },
+      "outputs": [],
+      "source": [
+        "B, T, d_m, N = 18, 32, 16, 8\n",
+        "\n",
+        "# run this to test if your code is running\n",
+        "def tranformer(sequences, num_heads=1, num_layers=1, widening_factor=4, d_m=d_m, return_weights=True):\n",
+        "  transform = Transformer(num_heads, num_layers, d_m, widening_factor)\n",
+        "  mask = jnp.tril(np.ones((sequences.shape[1], sequences.shape[1])))\n",
+        "  return transform(sequences, sequences, None, mask, return_weights)\n",
+        "\n",
+        "tranformer = hk.transform(tranformer)\n",
+        "\n",
+        "# initialise module and get dummy output\n",
+        "key = jax.random.PRNGKey(42)\n",
+        "X = jax.random.normal(key, [B, T,d_m])\n",
+        "params = tranformer.init(\n",
+        "    key, X, num_heads=1, num_layers=1, widening_factor=4, d_m=d_m, return_weights=True\n",
+        ")\n",
+        "\n",
+        "# extract output from decoder\n",
+        "logits, encoder_att_weights, decoder_att_weights = tranformer.apply(\n",
+        "    params, key, X, num_heads=1, num_layers=1, widening_factor=4, d_m=d_m, return_weights=True\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BLeXyXNE5qq9"
+      },
+      "source": [
+        "As a final sanity check, we can see that our attention weights behave as expected for now. The encoder weights can attend to all input sequences, and our decoder only attends to previous tokens."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "h0PNKe1r3O9H"
+      },
+      "outputs": [],
+      "source": [
+        "fig, ax = plt.subplots(1,2, figsize=(10,5))\n",
+        "plt.suptitle('Decoder attention weights (left) vs Encoder attention weights(right')\n",
+        "sns.heatmap(encoder_att_weights[0,0,0,...], ax=ax[0],cmap='Blues')\n",
+        "sns.heatmap(decoder_att_weights[0,0,0,...], ax=ax[1],cmap='Blues')\n",
+        "fig.show()"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "_bqOml7HYNmm"
+      },
+      "source": [
+        "### Training our model to invert sentence"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "M-f_jnA2P_k5"
+      },
+      "source": [
+        "To showcase the encoder-decoder transformer model we have just created, we will train it to reverse the order of an input string. For example:\n",
+        "\n",
+        "- *Input*: \"Look at me, I am reveresed.\"\n",
+        "- *Output*: \"reversed am I, look at me\"\n",
```

Comment:
Should this not be "reversed. am I me, at Look" instead?

Reply:
Nice catch thank you